File: python/src/aun_core/register_flow.py
from __future__ import annotations

import asyncio
import base64
import json
import re
import ssl
import time
from typing import Any
from urllib.parse import urlparse, urlunparse

import aiohttp
import websockets
from cryptography import x509
from cryptography.hazmat.primitives import serialization

from .cert_verifier import GatewayCertificateVerifier
from .crypto import CryptoProvider
from .errors import AuthError, ConnectionError, IdentityConflictError, ValidationError, map_remote_error
from .keystore.base import KeyStore


_AID_NAME_RE = re.compile(r'^[a-z0-9_][a-z0-9_-]{3,63}$')


class RegisterFlow:
    def __init__(
        self,
        keystore: KeyStore,
        crypto: CryptoProvider,
        *,
        verify_ssl: bool = False,
        root_ca_path: str | None = None,
        logger=None,
        net=None,
    ) -> None:
        self._keystore = keystore
        self._crypto = crypto
        self._verify_ssl = verify_ssl
        self._log = logger
        self._net = net
        self._certs = GatewayCertificateVerifier(
            root_ca_path=root_ca_path,
            store=keystore,
            verify_ssl=verify_ssl,
            logger=logger,
            net=net,
            module="register_flow",
        )

    def validate_aid_name(self, aid: str) -> None:
        self._validate_aid_name(aid)

    async def fetch_peer_cert(self, gateway_url: str, aid: str) -> str | None:
        return await self._download_registered_cert(gateway_url, aid)

    async def short_rpc(self, gateway_url: str, method: str, params: dict[str, Any]) -> dict[str, Any]:
        return await self._short_rpc(gateway_url, method, params)

    def generate_identity(self) -> dict[str, Any]:
        return self._crypto.generate_identity()

    def new_client_nonce(self) -> str:
        return self._crypto.new_client_nonce()

    async def verify_phase1_response(self, gateway_url: str, result: dict[str, Any], client_nonce: str) -> None:
        await self._certs.verify_phase1_response(gateway_url, result, client_nonce)

    def reload_trusted_roots(self) -> int:
        return self._certs.reload_trusted_roots()

    async def register_aid(self, gateway_url: str, aid: str) -> dict[str, Any]:
        """注册新 AID，返回含私钥字段的完整 dict，私钥由调用方（AIDStore）写入。"""
        _t_start = time.time()
        self._validate_aid_name(aid)
        self._debug("register_flow", "register_aid enter: aid=%s gateway=%s", aid, gateway_url)
        try:
            # Step 1: 本地已有 keypair → 查服务端做幂等/恢复
            existing = self._keystore.load_identity(aid)
            if existing and existing.get("private_key_pem") and existing.get("public_key_der_b64"):
                local_pub_der = base64.b64decode(existing["public_key_der_b64"])
                server_cert_pem = await self._download_registered_cert(gateway_url, aid)
                if server_cert_pem:
                    cert = x509.load_pem_x509_certificate(server_cert_pem.encode("utf-8"))
                    cert_pub_der = cert.public_key().public_bytes(
                        serialization.Encoding.DER,
                        serialization.PublicFormat.SubjectPublicKeyInfo,
                    )
                    if cert_pub_der != local_pub_der:
                        raise IdentityConflictError(
                            f"AID '{aid}' is registered by another party on server (public key mismatch). "
                            f"Choose a different name."
                        )
                    if not existing.get("cert"):
                        existing["cert"] = server_cert_pem
                        self._persist_identity(existing)
                    self._debug("register_flow", "register_aid exit (idempotent): elapsed=%.3fs aid=%s", time.time() - _t_start, aid)
                    return {"aid": aid, "cert": server_cert_pem, "private_key_pem": existing.get("private_key_pem", ""), "public_key_der_b64": existing.get("public_key_der_b64", ""), "curve": existing.get("curve", "P-256")}
                else:
                    created = await self._create_aid(gateway_url, existing)
                    cert_pem = created.get("cert", "")
                    if not cert_pem:
                        raise AuthError(f"register_aid: server response missing cert for {aid}")
                    existing["cert"] = cert_pem
                    cert = x509.load_pem_x509_certificate(cert_pem.encode("utf-8"))
                    cert_pub_der = cert.public_key().public_bytes(
                        serialization.Encoding.DER,
                        serialization.PublicFormat.SubjectPublicKeyInfo,
                    )
                    if cert_pub_der != local_pub_der:
                        raise AuthError(f"register_aid: server returned certificate with mismatched public key for {aid}")
                    self._persist_identity(existing)
                    self._debug("register_flow", "register_aid exit (recovered): elapsed=%.3fs aid=%s", time.time() - _t_start, aid)
                    return {"aid": aid, "cert": cert_pem, "private_key_pem": existing.get("private_key_pem", ""), "public_key_der_b64": existing.get("public_key_der_b64", ""), "curve": existing.get("curve", "P-256")}

            # Step 2: 检查 _pending/ 残留临时目录（崩溃恢复）
            recovered = await self._try_recover_pending_registration(gateway_url, aid)
            if recovered is not None:
                self._debug("register_flow", "register_aid recovered from pending: aid=%s", aid)
                return recovered

            # Step 3: 先查服务端确认未注册
            server_cert_pem = await self._download_registered_cert(gateway_url, aid)
            if server_cert_pem:
                raise IdentityConflictError(
                    f"AID '{aid}' is already registered on server. "
                    f"Choose a different name, or if you own the keypair use a recovery flow."
                )

            # Step 4: 创建 pending 目录 + 生成 keypair + 写入受保护私钥
            identity = self._crypto.generate_identity()
            identity["aid"] = aid
            pending_dir = self._keystore.pending_identity_dir(aid)
            self._keystore.save_pending_key_pair(pending_dir, aid, identity)

            # Step 5: RPC 注册
            try:
                created = await self._create_aid(gateway_url, identity)
            except Exception:
                self._debug("register_flow", "register_aid RPC failed; pending kept for recovery: aid=%s pending=%s", aid, pending_dir)
                raise
            cert_pem = created.get("cert", "")
            if not cert_pem:
                raise AuthError(f"register_aid: server response missing cert for {aid}")
            identity["cert"] = cert_pem

            # Step 6: 校验 cert 公钥 == 本地公钥
            cert = x509.load_pem_x509_certificate(cert_pem.encode("utf-8"))
            cert_pub_der = cert.public_key().public_bytes(
                serialization.Encoding.DER,
                serialization.PublicFormat.SubjectPublicKeyInfo,
            )
            local_pub_der = base64.b64decode(identity["public_key_der_b64"])
            if cert_pub_der != local_pub_der:
                raise AuthError(f"register_aid: server returned certificate with mismatched public key for {aid}")

            # Step 7: 写 cert 到 pending，Step 8: 原子 promote 到正式目录
            self._keystore.save_pending_cert(pending_dir, cert_pem)
            try:
                self._keystore.promote_pending_identity(pending_dir, aid)
            except FileExistsError as exc:
                raise IdentityConflictError(
                    f"AID '{aid}' was created by another process during registration; pending dir kept for cleanup."
                ) from exc

            # Step 9: 持久化（不含私钥，私钥已随 pending promote 到正式目录）
            self._persist_identity(identity)
            self._debug("register_flow", "register_aid exit: elapsed=%.3fs aid=%s", time.time() - _t_start, aid)
            return {"aid": aid, "cert": cert_pem, "private_key_pem": identity.get("private_key_pem", ""), "public_key_der_b64": identity.get("public_key_der_b64", ""), "curve": identity.get("curve", "P-256")}
        except Exception as exc:
            self._debug("register_flow", "register_aid exit (error): elapsed=%.3fs aid=%s err=%s", time.time() - _t_start, aid, exc)
            raise

    def _persist_identity(self, identity: dict[str, Any]) -> None:
        """持久化证书；私钥由 pending promote 或 AIDStore.save_key_pair 管理。"""
        aid = str(identity.get("aid") or "")
        if not aid:
            return
        cert_pem = str(identity.get("cert") or "")
        if cert_pem:
            self._keystore.save_cert(aid, cert_pem)

    async def _try_recover_pending_registration(self, gateway_url: str, aid: str) -> dict[str, Any] | None:
        for pending_dir in self._keystore.list_pending_identity_dirs(aid):
            private_data = self._keystore.load_pending_key_pair(pending_dir, aid)
            if private_data is None:
                self._keystore.discard_pending_identity(pending_dir)
                continue

            private_key_pem = str(private_data.get("private_key_pem") or "")
            public_key_der_b64 = str(private_data.get("public_key_der_b64") or "")
            if not private_key_pem or not public_key_der_b64:
                self._keystore.discard_pending_identity(pending_dir)
                continue

            server_cert_pem = await self._download_registered_cert(gateway_url, aid)
            if not server_cert_pem:
                self._debug("register_flow", "pending dir found but server has no registration; cleaning: %s", pending_dir)
                self._keystore.discard_pending_identity(pending_dir)
                return None

            cert = x509.load_pem_x509_certificate(server_cert_pem.encode("utf-8"))
            cert_pub_der = cert.public_key().public_bytes(
                serialization.Encoding.DER,
                serialization.PublicFormat.SubjectPublicKeyInfo,
            )
            local_pub_der = base64.b64decode(public_key_der_b64)
            if cert_pub_der != local_pub_der:
                self._keystore.discard_pending_identity(pending_dir)
                raise IdentityConflictError(
                    f"AID '{aid}' has been registered by another party while local pending registration was incomplete; "
                    f"local pending key discarded."
                )

            identity = {
                "aid": aid,
                "cert": server_cert_pem,
                "private_key_pem": private_key_pem,
                "public_key_der_b64": public_key_der_b64,
                "curve": private_data.get("curve", "P-256"),
            }
            self._keystore.save_pending_key_pair(pending_dir, aid, identity)
            self._keystore.save_pending_cert(pending_dir, server_cert_pem)
            try:
                self._keystore.promote_pending_identity(pending_dir, aid)
            except FileExistsError as exc:
                raise IdentityConflictError(
                    f"AID '{aid}' was created by another process during recovery; pending dir kept for cleanup."
                ) from exc
            self._persist_identity(identity)
            return identity
        return None

    async def _download_registered_cert(self, gateway_url: str, aid: str) -> str | None:
        cert_url = self._gateway_http_url(gateway_url, f"/pki/cert/{aid}")
        try:
            if self._net:
                cert_pem = await self._net.http_get_text(cert_url, timeout=5.0)
            else:
                timeout = aiohttp.ClientTimeout(total=5.0)
                ssl_param = None if self._verify_ssl else False
                async with aiohttp.ClientSession(timeout=timeout) as session:
                    async with session.get(cert_url, ssl=ssl_param) as response:
                        if response.status == 404:
                            return None
                        response.raise_for_status()
                        cert_pem = await response.text()
        except Exception as exc:
            if "404" in str(exc) or "Not Found" in str(exc):
                return None
            raise AuthError(f"failed to fetch {cert_url}") from exc
        if "BEGIN CERTIFICATE" not in cert_pem:
            raise AuthError(f"invalid certificate payload from {cert_url}")
        return cert_pem

    async def _create_aid(self, gateway_url: str, identity: dict[str, Any]) -> dict[str, Any]:
        response = await self._short_rpc(gateway_url, "auth.create_aid", {
            "aid": identity["aid"],
            "public_key": identity["public_key_der_b64"],
            "curve": identity.get("curve", "P-256"),
        })
        return {"cert": response["cert"]}

    async def _short_rpc(self, gateway_url: str, method: str, params: dict[str, Any]) -> dict[str, Any]:
        _SHORT_RPC_TIMEOUT = 15.0
        _CONNECT_TIMEOUT = 10.0
        try:
            ws = await asyncio.wait_for(self._connect(gateway_url), timeout=_CONNECT_TIMEOUT)
        except asyncio.TimeoutError:
            raise ConnectionError(f"gateway {gateway_url} WebSocket connect timeout ({_CONNECT_TIMEOUT}s)")
        try:
            try:
                await asyncio.wait_for(ws.recv(), timeout=_SHORT_RPC_TIMEOUT)
            except asyncio.TimeoutError:
                raise ConnectionError(f"gateway {gateway_url} handshake recv timeout ({_SHORT_RPC_TIMEOUT}s)")
            envelope = {"jsonrpc": "2.0", "id": f"pre-{method}", "method": method, "params": params}
            await ws.send(json.dumps(envelope, ensure_ascii=False, separators=(",", ":")))
            try:
                raw = await asyncio.wait_for(ws.recv(), timeout=_SHORT_RPC_TIMEOUT)
            except asyncio.TimeoutError:
                raise ConnectionError(f"gateway {gateway_url} RPC recv timeout for {method} ({_SHORT_RPC_TIMEOUT}s)")
            message = raw if isinstance(raw, dict) else json.loads(raw)
        finally:
            await ws.close()
        if "error" in message:
            raise map_remote_error(message["error"])
        result = message.get("result")
        if not isinstance(result, dict):
            raise ValidationError(f"invalid pre-auth response for {method}")
        if result.get("success") is False:
            raise AuthError(str(result.get("error", f"{method} failed")))
        return result

    async def _connect(self, url: str):
        kwargs: dict[str, Any] = {"open_timeout": 5, "close_timeout": 5, "ping_interval": None}
        if not self._verify_ssl and str(url).lower().startswith("wss://"):
            ssl_ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
            ssl_ctx.check_hostname = False
            ssl_ctx.verify_mode = ssl.CERT_NONE
            kwargs["ssl"] = ssl_ctx
        return await websockets.connect(url, **kwargs)

    @staticmethod
    def _gateway_http_url(gateway_url: str, path: str) -> str:
        parsed = urlparse(gateway_url)
        scheme = "https" if parsed.scheme == "wss" else "http"
        return urlunparse((scheme, parsed.netloc, path, "", "", ""))

    @staticmethod
    def _validate_aid_name(aid: str) -> None:
        name = aid.split(".")[0] if "." in aid else aid
        if not _AID_NAME_RE.match(name):
            raise ValidationError(
                f"Invalid AID name '{name}': must be 4-64 characters, "
                f"only [a-z0-9_-], cannot start with '-'"
            )
        if name.startswith("guest"):
            raise ValidationError("AID name must not start with 'guest'")

    def _debug(self, module: str, msg: str, *args: Any) -> None:
        if self._log is not None:
            self._log.debug(module, msg, *args)

File: js/src/register-flow.ts
/**
 * RegisterFlow — AID 注册流程。
 * 与 TS SDK register-flow.ts 对齐。浏览器环境使用原生 WebSocket + fetch。
 */

import type { KeyStore } from './keystore/index.js';
import type { CryptoProvider } from './crypto.js';
import type { ModuleLogger } from './logger.js';
import { AuthError, IdentityConflictError, ValidationError, mapRemoteError } from './errors.js';
import { isJsonObject, type IdentityRecord, type KeyPairRecord, type JsonObject, type RpcParams } from './types.js';
import { uint8ToBase64, pemToArrayBuffer } from './crypto.js';
import { parseCertMetadata, verifySignatureWithCert } from './cert-utils.js';

const _noopLog: ModuleLogger = { error: () => {}, warn: () => {}, info: () => {}, debug: () => {} };
const AID_NAME_RE = /^[a-z0-9_][a-z0-9_-]{3,63}$/;

type PendingKeyStore = KeyStore & {
  pendingIdentityDir(aid: string): Promise<string>;
  listPendingIdentityDirs(aid: string): Promise<string[]>;
  savePendingKeyPair(handle: string, aid: string, keyPair: KeyPairRecord): Promise<void>;
  loadPendingKeyPair(handle: string, aid: string): Promise<KeyPairRecord | null>;
  savePendingCert(handle: string, certPem: string): Promise<void>;
  promotePendingIdentity(handle: string, aid: string): Promise<string>;
  discardPendingIdentity?(handle: string): Promise<void>;
};

export interface RegisterResult {
  aid: string;
  cert: string;
  private_key_pem: string;
  public_key_der_b64: string;
  curve: string;
}

/** 将 WebSocket URL 转为对应的 HTTP URL */
function _gatewayHttpUrl(gatewayUrl: string, path: string): string {
  try {
    const parsed = new URL(gatewayUrl);
    const scheme = parsed.protocol === 'wss:' ? 'https:' : 'http:';
    return `${scheme}//${parsed.host}${path}`;
  } catch {
    return gatewayUrl.replace(/^wss:/, 'https:').replace(/^ws:/, 'http:') + path;
  }
}

/** 从证书 PEM 中提取 SPKI 公钥 DER base64（简化版，复用 auth.ts 的 parseCertDer 逻辑） */
function _extractSpkiB64FromPem(certPem: string): string {
  // 复用 auth.ts 中已有的 parseCertDer 逻辑，但 auth.ts 未导出。
  // 这里用最小化的 DER 解析提取 SPKI 字节。
  const der = new Uint8Array(pemToArrayBuffer(certPem));

  function readLen(data: Uint8Array, offset: number): { value: number; lenBytes: number } {
    const first = data[offset]!;
    if (!(first & 0x80)) return { value: first, lenBytes: 1 };
    const n = first & 0x7f;
    let v = 0;
    for (let i = 0; i < n; i++) v = (v << 8) | data[offset + 1 + i]!;
    return { value: v, lenBytes: 1 + n };
  }

  function skipTlv(data: Uint8Array, offset: number): number {
    const len = readLen(data, offset + 1);
    return offset + 1 + len.lenBytes + len.value;
  }

  // cert SEQUENCE → tbs SEQUENCE → skip version/serial/sigAlg/issuer/validity/subject → spki
  const certLen = readLen(der, 1);
  const tbsStart = 1 + certLen.lenBytes;
  const tbsLen = readLen(der, tbsStart + 1);
  let pos = tbsStart + 1 + tbsLen.lenBytes;
  if (der[pos] === 0xa0) pos = skipTlv(der, pos); // version
  pos = skipTlv(der, pos); // serial
  pos = skipTlv(der, pos); // sigAlg
  pos = skipTlv(der, pos); // issuer
  pos = skipTlv(der, pos); // validity
  pos = skipTlv(der, pos); // subject
  // now at spki SEQUENCE
  const spkiLen = readLen(der, pos + 1);
  const spkiEnd = pos + 1 + spkiLen.lenBytes + spkiLen.value;
  return uint8ToBase64(der.slice(pos, spkiEnd));
}

export class RegisterFlow {
  private _keystore: PendingKeyStore;
  private _crypto: CryptoProvider;
  private _logger: ModuleLogger;

  constructor(opts: {
    keystore: PendingKeyStore;
    crypto: CryptoProvider;
    verifySsl?: boolean;
    logger?: ModuleLogger;
  }) {
    this._keystore = opts.keystore;
    this._crypto = opts.crypto;
    this._logger = opts.logger ?? _noopLog;
    // verifySsl 在浏览器环境不可控，忽略
  }

  validateAidName(aid: string): void {
    RegisterFlow.validateAidName(aid);
  }

  static validateAidName(aid: string): void {
    const name = aid.includes('.') ? aid.split('.')[0]! : aid;
    if (!AID_NAME_RE.test(name)) {
      throw new ValidationError(
        `Invalid AID name '${name}': must be 4-64 characters, only [a-z0-9_-], cannot start with '-'`,
      );
    }
    if (name.startsWith('guest')) {
      throw new ValidationError("AID name must not start with 'guest'");
    }
  }

  async fetchPeerCert(gatewayUrl: string, aid: string): Promise<string | null> {
    return this._downloadRegisteredCert(gatewayUrl, aid);
  }

  async shortRpc(gatewayUrl: string, method: string, params: RpcParams): Promise<JsonObject> {
    return this._shortRpc(gatewayUrl, method, params);
  }

  async generateIdentity(): Promise<KeyPairRecord> {
    return await this._crypto.generateIdentity() as KeyPairRecord;
  }

  newClientNonce(): string {
    return this._crypto.newClientNonce();
  }

  async verifyPhase1Response(gatewayUrl: string, result: JsonObject, clientNonce: string): Promise<void> {
    void gatewayUrl;
    const authCertPem = String(result.auth_cert ?? '');
    const signatureB64 = String(result.client_nonce_signature ?? '');
    if (!authCertPem) throw new AuthError('aid_login1 missing auth_cert');
    if (!signatureB64) throw new AuthError('aid_login1 missing client_nonce_signature');

    const metadata = parseCertMetadata(authCertPem);
    const now = Date.now();
    if (metadata.notBefore.getTime() > now) throw new AuthError('aid_login1 auth certificate is not_yet_valid');
    if (metadata.notAfter.getTime() < now) throw new AuthError('aid_login1 auth certificate is expired');

    const valid = await verifySignatureWithCert(
      authCertPem,
      signatureB64,
      new TextEncoder().encode(clientNonce),
    );
    if (!valid) throw new AuthError('aid_login1 server auth signature verification failed');
  }

  reloadTrustedRoots(): number {
    return 0;
  }

  async registerAid(gatewayUrl: string, aid: string): Promise<RegisterResult> {
    const tStart = Date.now();
    this.validateAidName(aid);
    this._logger.debug(`registerAid enter: aid=${aid}, gateway=${gatewayUrl}`);
    try {
      // Step 1: 本地已有 keypair → 幂等/恢复
      const existing = await this._keystore.loadIdentity(aid);
      if (existing && existing.private_key_pem && existing.public_key_der_b64) {
        this._logger.debug(`registerAid: local keypair exists, checking server: aid=${aid}`);
        const localPubB64 = String(existing.public_key_der_b64);
        const serverCertPem = await this._downloadRegisteredCert(gatewayUrl, aid);
        if (serverCertPem) {
          // 服务端已注册 → 比对公钥
          const serverPubB64 = _extractSpkiB64FromPem(serverCertPem);
          if (serverPubB64 !== localPubB64) {
            throw new IdentityConflictError(
              `AID '${aid}' is registered by another party on server (public key mismatch). Choose a different name.`,
            );
          }
          this._logger.info(`registerAid: idempotent return for already-registered AID: aid=${aid}`);
          if (!existing.cert) {
            existing.cert = serverCertPem;
            await this._persistIdentity(existing);
          }
          this._logger.debug(`registerAid exit (idempotent): elapsed=${Date.now() - tStart}ms aid=${aid}`);
          return {
            aid,
            cert: serverCertPem,
            private_key_pem: String(existing.private_key_pem),
            public_key_der_b64: localPubB64,
            curve: String(existing.curve ?? 'P-256'),
          };
        } else {
          // 服务端无记录 → 用现有 keypair 发起注册
          this._logger.debug(`registerAid: server has no record, registering with existing keypair: aid=${aid}`);
          const created = await this._createAid(gatewayUrl, existing);
          const certPem = String(created.cert ?? '');
          if (!certPem) throw new AuthError(`registerAid: server response missing cert for ${aid}`);
          existing.cert = certPem;
          this._assertCertMatchesLocalKeypair(existing);
          await this._persistIdentity(existing);
          this._logger.debug(`registerAid exit (recovered): elapsed=${Date.now() - tStart}ms aid=${aid}`);
          return {
            aid,
            cert: certPem,
            private_key_pem: String(existing.private_key_pem),
            public_key_der_b64: localPubB64,
            curve: String(existing.curve ?? 'P-256'),
          };
        }
      }

      // Step 2: 检查 pending 残留（崩溃/页面关闭恢复）
      const recovered = await this._tryRecoverPendingRegistration(gatewayUrl, aid);
      if (recovered !== null) {
        this._logger.info(`registerAid recovered from pending: aid=${aid}`);
        return recovered;
      }

      // Step 3: 服务端查重
      const existingCert = await this._downloadRegisteredCert(gatewayUrl, aid);
      if (existingCert) {
        throw new IdentityConflictError(
          `AID '${aid}' is already registered on server. Choose a different name, or if you own the keypair use a recovery flow.`,
        );
      }

      // Step 4: 生成 keypair，并先写入 pending 加密记录
      const identity = await this._crypto.generateIdentity() as IdentityRecord;
      identity.aid = aid;
      const pendingStore = this._pendingStore();
      const pendingHandle = await pendingStore.pendingIdentityDir(aid);
      await pendingStore.savePendingKeyPair(pendingHandle, aid, identity);

      // Step 5: 服务端注册拿 cert；失败时保留 pending 供后续恢复
      let created: JsonObject;
      try {
        created = await this._createAid(gatewayUrl, identity);
      } catch (e) {
        this._logger.warn(`registerAid RPC failed (pending kept for recovery): aid=${aid}, pending=${pendingHandle}, error=${e instanceof Error ? e.message : String(e)}`);
        throw e;
      }
      const certPem = String(created.cert ?? '');
      if (!certPem) throw new AuthError(`registerAid: server response missing cert for ${aid}`);
      identity.cert = certPem;

      // Step 6: 校验 cert 公钥 == 本地公钥
      this._assertCertMatchesLocalKeypair(identity);

      // Step 7/8: 写 pending cert，再 promote 到正式 identity
      await pendingStore.savePendingCert(pendingHandle, certPem);
      try {
        await pendingStore.promotePendingIdentity(pendingHandle, aid);
      } catch (e) {
        throw new IdentityConflictError(
          `AID '${aid}' was created by another process during registration; pending record kept for cleanup.`,
        );
      }

      // Step 9: 持久化非私钥字段；私钥已随 pending promote 转正
      await this._persistIdentity(identity);
      this._logger.debug(`registerAid exit: elapsed=${Date.now() - tStart}ms aid=${aid}`);
      return {
        aid,
        cert: certPem,
        private_key_pem: String(identity.private_key_pem ?? ''),
        public_key_der_b64: String(identity.public_key_der_b64 ?? ''),
        curve: String(identity.curve ?? 'P-256'),
      };
    } catch (err) {
      this._logger.debug(`registerAid exit (error): elapsed=${Date.now() - tStart}ms err=${err instanceof Error ? err.message : String(err)}`);
      throw err;
    }
  }

  private _pendingStore(): PendingKeyStore {
    const store = this._keystore as Partial<PendingKeyStore>;
    for (const name of [
      'pendingIdentityDir',
      'listPendingIdentityDirs',
      'savePendingKeyPair',
      'loadPendingKeyPair',
      'savePendingCert',
      'promotePendingIdentity',
    ] as const) {
      if (typeof store[name] !== 'function') {
        throw new AuthError(`keystore does not support pending registration: ${name}`);
      }
    }
    return store as PendingKeyStore;
  }

  private async _tryRecoverPendingRegistration(gatewayUrl: string, aid: string): Promise<RegisterResult | null> {
    const pendingStore = this._pendingStore();
    const handles = await pendingStore.listPendingIdentityDirs(aid);
    for (const handle of handles) {
      let keyPair: IdentityRecord | null;
      try {
        keyPair = await pendingStore.loadPendingKeyPair(handle, aid);
      } catch (exc) {
        this._logger.warn(`pending identity keypair load failed; keeping pending for retry: aid=${aid}, pending=${handle}, error=${exc instanceof Error ? exc.message : String(exc)}`);
        throw exc;
      }
      const privateKeyPem = String(keyPair?.private_key_pem ?? '');
      const publicKeyDerB64 = String(keyPair?.public_key_der_b64 ?? '');
      if (!privateKeyPem || !publicKeyDerB64) {
        await pendingStore.discardPendingIdentity?.(handle);
        continue;
      }

      const serverCertPem = await this._downloadRegisteredCert(gatewayUrl, aid);
      if (!serverCertPem) {
        this._logger.info(`pending record found but server has no registration; cleaning up: ${handle}`);
        await pendingStore.discardPendingIdentity?.(handle);
        return null;
      }

      const serverPubB64 = _extractSpkiB64FromPem(serverCertPem);
      if (serverPubB64 !== publicKeyDerB64) {
        await pendingStore.discardPendingIdentity?.(handle);
        throw new IdentityConflictError(
          `AID '${aid}' has been registered by another party while local pending registration was incomplete; local pending key discarded.`,
        );
      }

      const identity: IdentityRecord = {
        aid,
        cert: serverCertPem,
        private_key_pem: privateKeyPem,
        public_key_der_b64: publicKeyDerB64,
        curve: String(keyPair?.curve ?? 'P-256'),
      };
      await pendingStore.savePendingKeyPair(handle, aid, identity);
      await pendingStore.savePendingCert(handle, serverCertPem);
      try {
        await pendingStore.promotePendingIdentity(handle, aid);
      } catch (e) {
        throw new IdentityConflictError(
          `AID '${aid}' was created by another process during recovery; pending record kept for cleanup.`,
        );
      }
      await this._persistIdentity(identity);
      return {
        aid,
        cert: serverCertPem,
        private_key_pem: privateKeyPem,
        public_key_der_b64: publicKeyDerB64,
        curve: String(keyPair?.curve ?? 'P-256'),
      };
    }
    return null;
  }

  private async _downloadRegisteredCert(gatewayUrl: string, aid: string): Promise<string | null> {
    const certUrl = _gatewayHttpUrl(gatewayUrl, `/pki/cert/${aid}`);
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), 5000);
    try {
      const resp = await fetch(certUrl, { signal: controller.signal });
      if (!resp.ok) return null;
      const certPem = await resp.text();
      if (!certPem || !certPem.includes('BEGIN CERTIFICATE')) return null;
      return certPem;
    } catch {
      return null;
    } finally {
      clearTimeout(timer);
    }
  }

  private async _createAid(gatewayUrl: string, identity: IdentityRecord): Promise<JsonObject> {
    const response = await this._shortRpc(gatewayUrl, 'auth.create_aid', {
      aid: identity.aid,
      public_key: identity.public_key_der_b64,
      curve: identity.curve ?? 'P-256',
    });
    return { cert: response.cert };
  }

  private _shortRpc(gatewayUrl: string, method: string, params: RpcParams): Promise<JsonObject> {
    return new Promise((resolve, reject) => {
      let ws: WebSocket;
      try {
        ws = new WebSocket(gatewayUrl);
      } catch {
        reject(new AuthError(`WebSocket 连接失败: ${gatewayUrl}`));
        return;
      }
      let receivedChallenge = false;
      const timeout = globalThis.setTimeout(() => {
        try { ws.close(); } catch { /* 忽略 */ }
        reject(new AuthError(`shortRpc 超时: ${method}`));
      }, 10000);

      ws.onerror = () => {
        globalThis.clearTimeout(timeout);
        reject(new AuthError(`WebSocket 连接错误: ${gatewayUrl}`));
      };

      ws.onmessage = (event) => {
        try {
          const msg = typeof event.data === 'string' ? JSON.parse(event.data) : event.data;
          if (!receivedChallenge) {
            receivedChallenge = true;
            ws.send(JSON.stringify({ jsonrpc: '2.0', id: `pre-${method}`, method, params }));
            return;
          }
          globalThis.clearTimeout(timeout);
          try { ws.close(); } catch { /* 忽略 */ }
          if (msg.error) { reject(mapRemoteError(msg.error)); return; }
          const result = msg.result;
          if (!isJsonObject(result)) { reject(new ValidationError(`invalid pre-auth response for ${method}`)); return; }
          if (result.success === false) { reject(new AuthError(String(result.error ?? `${method} failed`))); return; }
          resolve(result);
        } catch (e) {
          globalThis.clearTimeout(timeout);
          try { ws.close(); } catch { /* 忽略 */ }
          reject(e instanceof Error ? e : new AuthError(String(e)));
        }
      };
    });
  }

  private _assertCertMatchesLocalKeypair(identity: IdentityRecord): void {
    const aid = identity.aid ?? '?';
    const certPem = identity.cert as string | undefined;
    const localPubB64 = identity.public_key_der_b64 as string | undefined;
    if (!certPem || !localPubB64) {
      throw new AuthError(`identity for aid ${aid} missing cert or public key`);
    }
    const certPubB64 = _extractSpkiB64FromPem(certPem);
    if (certPubB64 !== localPubB64) {
      throw new AuthError(`local certificate public key does not match local keypair for aid ${aid}`);
    }
  }

  private async _persistIdentity(identity: IdentityRecord): Promise<void> {
    const aid = String(identity.aid ?? '');
    if (!aid) return;
    const certPem = String(identity.cert ?? '');
    if (certPem) await this._keystore.saveCert(aid, certPem);
  }
}
